### PR TITLE
Add support for writer version 6 without DML in Delta Lake

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Relates to #12640
`INSERT`, `DELETE`, `UPDATE` and `MERGE` statements are still unsupported. 

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Allow metadata changes on Delta Lake tables with [identity columns](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#identity-columns). ({issue}`18200`)
```

### CODE REVIEW

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+                    .hasMessageMatching(".* Table .* requires Delta Lake writer version 7 which is not supported");
+        }
+        finally {
+            dropDeltaTableWithRetry("default." + tableName);
+        }
+    }
+
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_EXCLUDE_73, DELTA_LAKE_EXCLUDE_91, PROFILE_SPECIFIC_TESTS})
+    @Flaky(issue = DATABRICKS_COMMUNICATION_FAILURE_ISSUE, match = DATABRICKS_COMMUNICATION_FAILURE_MATCH)
+    public void testWritesToTableWithIdentityColumnFails()
```
Comment (homar) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java:106]: What is the difference between this test and the test above?

**Review by homar — APPROVED** [review]

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+import static io.trino.tests.product.TestGroups.PROFILE_SPECIFIC_TESTS;
+import static io.trino.tests.product.deltalake.util.DeltaLakeTestUtils.DATABRICKS_COMMUNICATION_FAILURE_ISSUE;
+import static io.trino.tests.product.deltalake.util.DeltaLakeTestUtils.DATABRICKS_COMMUNICATION_FAILURE_MATCH;
+import static io.trino.tests.product.deltalake.util.DeltaLakeTestUtils.dropDeltaTableWithRetry;
+import static io.trino.tests.product.utils.QueryExecutors.onDelta;
+import static io.trino.tests.product.utils.QueryExecutors.onTrino;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class TestDeltaLakeIdentityColumnCompatibility
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java:39]: > Extract column identity product test about Delta Lake

Any particular reason for this change?
Maybe something like

```
Extract Delta identity column product test to own class

It was in TestDeltaLakeAlterTableCompatibility, but that's incorrect
as the test isn't centered around ALTER TABLE.
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java [hunk 3]
```diff
@@ -52,30 +56,269 @@ public void testIdentityColumn()
...
-            assertQueryFailure(() -> onTrino().executeQuery("ALTER TABLE delta.default." + tableName + " ADD COLUMN c INT"))
-                    .hasMessageMatching(failureMessage);
-            assertQueryFailure(() -> onTrino().executeQuery("INSERT INTO delta.default." + tableName + "(a) VALUES (0)"))
-                    .hasMessageMatching(failureMessage);
-            assertQueryFailure(() -> onTrino().executeQuery("DELETE FROM delta.default." + tableName))
-                    .hasMessageMatching(failureMessage);
-            assertQueryFailure(() -> onTrino().executeQuery("MERGE INTO delta.default." + tableName + " t USING delta.default." + tableName + " s " + "ON (t.a = s.a) WHEN MATCHED THEN UPDATE SET b = 1"))
-                    .hasMessageMatching(failureMessage);
+            onTrino().executeQuery("COMMENT ON COLUMN delta.default." + tableName + ".b IS 'test column comment'");
+            onTrino().executeQuery("COMMENT ON TABLE delta.default." + tableName + " IS 'test table comment'");
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java:60]: verify the comment was persisted?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java [hunk 4]
```diff
@@ -52,30 +56,269 @@ public void testIdentityColumn()
...
             assertThat(onDelta().executeQuery("SELECT * FROM default." + tableName)).containsOnly(expected);
         }
         finally {
             dropDeltaTableWithRetry("default." + tableName);
         }
     }
+
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_EXCLUDE_73, DELTA_LAKE_EXCLUDE_91, DELTA_LAKE_EXCLUDE_104, DELTA_LAKE_EXCLUDE_113, PROFILE_SPECIFIC_TESTS})
+    @Flaky(issue = DATABRICKS_COMMUNICATION_FAILURE_ISSUE, match = DATABRICKS_COMMUNICATION_FAILURE_MATCH)
+    public void testIdentityColumnFeature()
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java:79]: the other method is called `testIdentityColumn`.
"Feature" suffix doesn't add any information, so it feels like "testIdentityColumn2".

can we name this method differently?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java [hunk 5]
```diff
@@ -52,30 +56,269 @@ public void testIdentityColumn()
...
+            onTrino().executeQuery("COMMENT ON COLUMN delta.default." + tableName + ".uppercase_identity IS 'test comment'");
+            onTrino().executeQuery("ALTER TABLE delta.default." + tableName + " DROP COLUMN renamed_col");
+
+            assertThat((String) onDelta().executeQuery("SHOW CREATE TABLE default." + tableName).getOnlyValue())
+                    .contains("UPPERCASE_IDENTITY BIGINT GENERATED ALWAYS AS IDENTITY");
+
+            // Verify the connector preserves column identity property when renaming columns containing uppercase characters
+            onTrino().executeQuery("ALTER TABLE delta.default." + tableName + " RENAME COLUMN uppercase_identity TO renamed_identity");
+            assertThat((String) onDelta().executeQuery("SHOW CREATE TABLE default." + tableName).getOnlyValue())
+                    .contains("renamed_identity BIGINT GENERATED ALWAYS AS IDENTITY");
```
Comment (findepi) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java:174]: There are some more fine-grained identity properties like delta.identity.start or delta.identity.step.
Let's have a test that these are not lost when we do metadata changes on a table, or when we write a checkpoint file.

**Comment by findepi** [comment]
can you please add explicit checkpoint verifications? the checkpoint file writer needs not to lose columns' identity properties

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java (hunk 1 above)
Comment (findinpath) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java:106]: In the above test, https://github.com/delta-io/delta/blob/master/PROTOCOL.md#valid-feature-names-in-table-features the protocol version points the min writer version to be `7`

```
{
  "protocol": {
    "minReaderVersion": 1,
    "minWriterVersion": 7,
```

while in this test, the inferred writer version will be `6`.

**Review by findinpath — APPROVED** [review]

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java (hunk 4 above)
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeIdentityColumnCompatibility.java:79]: The "Feature" comes from [table **features**](https://github.com/delta-io/delta/blob/master/PROTOCOL.md#table-features) in Delta. Renamed to `testIdentityColumnTableFeature`

**Review by findepi — APPROVED** [review]

**Pull request merged**